File: 1494.并行课程-ii.cpp
/*
 * @lc app=leetcode.cn id=1494 lang=cpp
 *
 * [1494] 并行课程 II
 */

// @lc code=start
#include<vector>
#include<climits>
#include<iostream>
#include<unordered_map>
#include<bitset>
using namespace std;
class Solution {
public:
    vector<vector<int>> adjlist;
    unordered_map<int, int> cache;
    int K;
    int hammingWeight(uint32_t n) {
        n = n - ((n >> 1) & 0x55555555);
        n = (n & 0x33333333) + ((n >> 2) & 0x33333333);
        n = (n + (n >> 4)) & 0x0f0f0f0f;
        n = n + (n >> 8);
        n = n + (n >> 16);
        return n & 0x3f;
    }

    vector<vector<int>> combinations(const vector<int>& arr, int k) {
        vector<vector<int>> ret;
        for (int i = 0; i < (1 << arr.size()); ++i) {
            if (hammingWeight(i) != k) continue;
            ret.push_back({});
            int tmp = i, idx = 0;
            while (tmp) {
                if (tmp & 1) ret.back().push_back(arr[idx]);
                ++idx;
                tmp >>= 1;
            }
        }
        return ret;
    }

    int dfs(vector<int>& in_degrees, int vis) {
        if (cache.count(vis)) return cache[vis];
        return cache[vis] = __dfs(in_degrees, vis);
    }

    int __dfs(vector<int>& in_degrees, int vis) {
        if (vis == 0) return 0;
        vector<int> starts;
        for (int i = 0; i < in_degrees.size(); ++i) {
            if (in_degrees[i] == 0 && (vis & (1 << i))) starts.push_back(i);
        }
        int ret = INT_MAX;
        for (auto& candis : combinations(starts, min((int)starts.size(), K))) {
            for (auto cur : candis) {
                vis &= ~(1 << cur);
                for (auto neibour : adjlist[cur]) {
                    --in_degrees[neibour];
                }
            }
            ret = min(ret, dfs(in_degrees, vis));
            for (auto cur : candis) {
                vis |= (1 << cur);
                for (auto neibour : adjlist[cur]) {
                    ++in_degrees[neibour];
                }
            }
        }
        return ret + 1;
    }
    int minNumberOfSemesters(int n, const vector<vector<int>>& relations, int k) {
        adjlist.resize(n, {});
        cache.clear();
        K = k;
        vector<int> in_degrees(n, 0);
        for (auto& link : relations) {
            adjlist[link[0] - 1].push_back(link[1] - 1);
            ++in_degrees[link[1] - 1];
        }
        return dfs(in_degrees, (1 << n) - 1);
    }
};
// @lc code=end

File: 1494.并行课程-ii.py
#
# @lc app=leetcode.cn id=1494 lang=python3
#
# [1494] 并行课程 II
#

# @lc code=start
from typing import List, Dict
from itertools import combinations
from math import inf

def cache(d: Dict):
    def wrapper(func):
        def myfunc(self, in_degrees: List[int], vis: int):
            key = vis
            if key in d: return d[key]
            val = func(self, in_degrees, vis)
            d[key] = val
            return val
        return myfunc
    return wrapper

class Solution:
    d = {}
    @cache(d)
    def dfs(self, in_degrees: List[int], vis: int) -> int:
        if vis == 0: return 0
        starts = [i for i, deg in enumerate(in_degrees) if deg == 0 and vis & (1<<i)]
        ret = inf
        for candis in combinations(starts, min(len(starts), self.k)):
            for cur in candis:
                vis &= ~(1<<cur)
                for neibour in self.adjlist[cur]:
                    in_degrees[neibour] -= 1
            ret = min(ret, self.dfs(in_degrees, vis))
            for cur in candis:
                vis |= (1<<cur)
                for neibour in self.adjlist[cur]:
                    in_degrees[neibour] += 1
        return ret + 1

    def minNumberOfSemesters(self, n: int, relations: List[List[int]], k: int) -> int:
        self.d.clear()
        self.adjlist = adjlist = [[] for _ in range(n)]
        self.k = k
        in_degrees = [0] * n
        for s, t in relations:
            adjlist[s-1].append(t-1)
            in_degrees[t-1] += 1
        
        return self.dfs(in_degrees, (1<<n)-1)


# @lc code=end

print(Solution().minNumberOfSemesters(n = 4, relations = [[2,1],[3,1],[1,4]], k = 2))
print(Solution().minNumberOfSemesters(n = 5, relations = [[2,1],[3,1],[4,1],[1,5]], k = 2))
print(Solution().minNumberOfSemesters(n = 11, relations = [], k = 2))
print(Solution().minNumberOfSemesters(13, [[12,8],[2,4],[3,7],[6,8],[11,8],[9,4],[9,7],[12,4],[11,4],[6,4],[1,4],[10,7],[10,4],[1,7],[1,8],[2,7],[8,4],[10,8],[12,7],[5,4],[3,4],[11,7],[7,4],[13,4],[9,8],[13,8]], 9))
print(Solution().minNumberOfSemesters(2, [], 1))